src/ArborX_BruteForce.hpp

```diff
@@ -123,21 +121,51 @@ class BruteForce
   void query(ExecutionSpace const &space, Predicates const &predicates,
              Callback const &callback, Ignore = Ignore()) const
   {
+    Details::check_valid_callback<int>(callback, predicates);
     base_type::query(space, predicates,
-                     Details::LegacyCallbackWrapper<
-                         Callback, typename base_type::value_type>{callback});
+                     Details::LegacyCallbackWrapper{callback});
   }
 
-  template <typename ExecutionSpace, typename Predicates,
-            typename CallbackOrView, typename View, typename... Args>
+  template <typename ExecutionSpace, typename Predicates, typename View,
+            typename... Args>
   std::enable_if_t<Kokkos::is_view_v<std::decay_t<View>>>
-  query(ExecutionSpace const &space, Predicates const &predicates,
-        CallbackOrView &&callback_or_view, View &&view, Args &&...args) const
+  query(ExecutionSpace const &space, Predicates const &predicates, View &&view,
+        Args &&...args) const
   {
-    base_type::query(space, predicates,
-                     std::forward<CallbackOrView>(callback_or_view),
+    base_type::query(space, predicates, Details::DefaultLegacyCallback{},
                      std::forward<View>(view), std::forward<Args>(args)...);
   }
+
+  template <typename ExecutionSpace, typename Predicates, typename Callback,
+            typename OutputView, typename OffsetView, typename... Args>
+  std::enable_if_t<!Kokkos::is_view_v<std::decay_t<Callback>>>
+  query(ExecutionSpace const &space, Predicates const &predicates,
+        Callback &&callback, OutputView &&out, OffsetView &&offset,
+        Args &&...args) const
+  {
+    if constexpr (!Details::is_tagged_post_callback<
+                      std::decay_t<Callback>>::value)
+    {
+      Details::check_valid_callback<int>(callback, predicates, out);
+      base_type::query(
+          space, predicates,
+          Details::LegacyCallbackWrapper{std::forward<Callback>(callback)},
+          std::forward<OutputView>(out), std::forward<OffsetView>(offset),
+          std::forward<Args>(args)...);
+    }
+    else
+    {
+      KokkosExt::ScopedProfileRegion guard("ArborX::BruteForce::query_crs");
+
+      Kokkos::View<int *, MemorySpace> indices(
+          "ArborX::CrsGraphWrapper::query::indices", 0);
+      base_type::query(space, predicates, Details::DefaultLegacyCallback{},
+                       indices, std::forward<OffsetView>(offset),
+                       std::forward<Args>(args)...);
+      callback(predicates, std::forward<OffsetView>(offset), indices,
+               std::forward<OutputView>(out));
+    }
+  }
```

Comment:
I moved the calls for the legacy classes out of `CrsGraphImpl`. The reason being is that the call stack was too complicated to figure out from within `CrsGraphImpl` when to wrap and when not to. Moving out made things doable.

The downside is that it had to be repeated in both `BruteForce` and `Testing::LegacyTree`. I think it's ok.